server/client.go

```diff
@@ -1966,6 +2001,65 @@ func (c *client) canSubscribe(subject string) bool {
 	return allowed
 }
 
+func (c *client) canQueueSubscribe(subject, queue string) bool {
+	if c.perms == nil {
+		return true
+	}
+
+	allowed := true
+	if c.perms.sub.allow != nil {
+		r := c.perms.sub.allow.Match(subject)
+
+		// If there is an explicit sub permission it allows any
+		// queue subscription unless there are explicit rules
+		// involving queue subscriptions.
+		subAllowed := len(r.psubs) != 0
+		allQueuesAllowed := len(r.qsubs) == 0
+		allowed = subAllowed && allQueuesAllowed
+
+		// We first check for explicit permissions for queue subscriptions,
+		// we later do another check for all subscriptions instead.
+		for _, qsub := range r.qsubs {
+			qq := qsub[0]
+			qname := string(qq.queue)
+			// NOTE: '*' and '>' tokens can also be valid
+			// queue names so we first check against the
+			// literal name.  e.g. v1.* == v1.*
+			if queue == qname || subjectHasWildcard(qname) && subjectIsSubsetMatch(queue, qname) {
+				allowed = true
+				break
+			}
+		}
+	}
+
+	// Check whether there is an explicit deny against this queue subscription.
+	if allowed && c.perms.sub.deny != nil {
+		r := c.perms.sub.deny.Match(subject)
+		subAllowed := len(r.psubs) == 0
+		allQueuesAllowed := len(r.qsubs) == 0
+
+		if subAllowed && !allQueuesAllowed {
+			for _, qsub := range r.qsubs {
```

Comment:
This could be another helper to check whether qname matches both in allow/deny logic.